modules/service/src/main/resources/db/migration/V0855__program_reference_update2.sql

```diff
@@ -0,0 +1,187 @@
+-- Describe the new types in t_program_type
+INSERT INTO t_program_type VALUES ('commissioning', 'COM', 'Commissioning', true);
+INSERT INTO t_program_type VALUES ('monitoring',    'MON', 'Monitoring',    true);
+
+-- Update the table constraint on reference types.
+ALTER TABLE t_program
+  DROP CONSTRAINT reference_type_constraint;
+
+ALTER TABLE t_program
+  ADD CONSTRAINT reference_type_constraint CHECK (
+      CASE
+          WHEN c_program_type = 'calibration'   OR
+               c_program_type = 'commissioning' OR
+               c_program_type = 'engineering'   OR
+               c_program_type = 'monitoring'    THEN
+            c_instrument      IS NOT NULL AND
+            c_library_desc    IS NULL     AND
+            c_semester        IS NOT NULL AND
+            c_semester_index  IS NOT NULL AND
+            c_science_subtype IS NULL
+
+          WHEN c_program_type = 'example' THEN
+            c_instrument      IS NOT NULL AND
+            c_library_desc    IS NULL     AND
+            c_semester        IS NULL     AND
+            c_semester_index  IS NULL     AND
+            c_science_subtype IS NULL
+
+          WHEN c_program_type = 'library' THEN
+            c_instrument      IS NOT NULL AND
+            c_library_desc    IS NOT NULL AND
+            c_semester        IS NULL     AND
+            c_semester_index  IS NULL     AND
+            c_science_subtype IS NULL
+
+          WHEN c_program_type = 'science' THEN
+            c_instrument   IS NULL     AND
+            c_library_desc IS NULL     AND
+            (c_proposal_status = 'not_submitted' OR (
+              c_semester        IS NOT NULL AND
+              c_semester_index  IS NOT NULL AND
+              c_science_subtype IS NOT NULL
+            ))
+      END
+  );
+
+-- Update the next_semester_index function so that COM and MON are handled.
+CREATE OR REPLACE FUNCTION next_semester_index(
+  program_type e_program_type,
+  semester     d_semester,
+  instrument   d_tag
+)
+RETURNS INT AS $$
+DECLARE
+    abbr text;
+    prefix text;
+    sequence_name VARCHAR;
+BEGIN
+    SELECT c_abbr INTO abbr FROM t_program_type WHERE c_type = program_type;
+    prefix := CONCAT('s_', semester, '_', abbr);
+
+    IF instrument IS NOT NULL THEN
+        sequence_name := CONCAT(prefix, '_', instrument);
+    ELSE
```

Comment:
Updating a function requires, as far as I know, completely rewriting it.  Old migrations cannot be touched.  The relevant change is that now it looks for whether `instrument` is `NULL` when deciding to include the instrument in the sequence name.  For `CAL`, `COM`, `ENG` and `MON` it will be supplied.  For science it will be `NULL`.